rfcs/007-background-workers.md

```diff
@@ -0,0 +1,299 @@
+
+### Clean Worker Shutdown
+
+If a worker is stopped cleanly, they MUST stop accepting new jobs from the
+pending queue, MUST stop running their jobs, MUST resubmit those jobs
+back into the pending queue. Finally, they MUST clear their health entry and
+remove themselves from workers membership list.
+
+### User Job Cancellation
+
+If a user decides to a cancel an active job, they MUST update the `CancelReq`
+field to `true`. Workers should periodically monitor that field and should stop
+running the job as soon as they detect it becoming `true`.
+
+## `/_active_tasks` API Compatibility
+
+The shape of the active table was deliberately constructed with the hopes of
+using it to implement the `/_active_tasks` API endpoint.
+
+When the HTTP API is accessed, a single `get_range(...)` request to
+`("couch_workers", "active")` subspace MAY be used to generate all the
+`_active_tasks` results. A job's `JobOpts` and `JobInfo` fields should contain
+most of needed information.
+
+# Advantages and Disadvantages
+
+Another alternative discussed on the mailing list is to mimic the
+pre-FoundationDB replication job scheduling model, where replication jobs are
+distributed between all the workers immediately when they are created. Each
+worker has their own separate job queue. The advantage of that model is that
+the queues and jobs are sharded evenly amongst the workers, avoiding potential
+hot key ranges around the global queue pushes and pops. The disadvantage is that it
+seemed like a premature optimization. Also, it would only allow for a completely
+fair scheduling algorithm as opposed letting workers throttle the number of
+jobs they pick up from the global queue based on load or other scheduling
+parameters.
+
+An alternative to health monitoring of individual workers is to periodically
+sort jobs based on a `LastUpdated` timestamp field and then automatically
+re-enqueue jobs that have not been updated in a while. The disadvantage of that
+approach is that it even though it might simplify the data model, it doesn't
+eliminate the need for having workers. Since workers are already present, it
+might be useful to expose their state explicitly so it can be inspected and
+monitored.
+
+## Possible Future Extensions
+
+Since all job keys and values are just FDB tuples and json encoded objects, in
+the future it might be possible to accept external jobs, not just jobs defined
+by the CouchDB internals. Also, since workers could be written in any language
+as long as they can talk to the FDB cluster, and follow the behavior describes
+in the design, it opens the possibility in the future have custom (user
+defined) workers of different types.
+
+# Key Changes
+
+ - New job execution framework
+ - A single global job queue for each job type
+ - Workers can dynamically scale up or down
+ - Worker health monitoring to detect dead workers
+ - Potential `_active_tasks` API implementation
+
+## Applications and Modules affected
+
+Replication and indexing
```

Comment:
+ couch-peruser, and any other background/cleanup jobs we may or may not be introducing with the new data models.

Reply:
Good call.